### ISSUE TITLE
[5.0.100-rtm.20508.12] TuringTrader Single File app Load Algorithms failed

### ISSUE THREAD

**Issue by jiangzeng01** [dotnet/runtime#43251]
**Application Name:** TuringTrader
**OS:** WIN10 RS5
**CPU:** amd64
**.NET SDK version:** 5.0.100-rtm.20508.12
**Application Location:**
\\\appcompatsrv\Appcompat\DotNetCore\AppSource\SingleFileP0AppsTarget5\Apps\TuringTrader\TuringTrader.exe
**Application Source Location:**
\\\appcompatsrv\Appcompat\DotNetCore\AppSource\SingleFileP0AppsTarget5\AppSources\TuringTrader
**DevDiv bug:** [https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1228888](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1228888)

 **Repro steps:**
The app was re-tartgeted to .NET 5.0 and published with command "dotnet publish -r win-x64 -p:PublishSingleFile=true --self-contained true -p:IncludeAllContentForSelfExtract=true"
Copy TuringTrader to local machine %APP_CLIENT_ROOT%.
1.Launch TuringTrader.exe  from %APP_CLIENT_ROOT%\TuringTrader folder.
2.Max up the application.
3.Click “File” menu.
4. Select "Algorithm" option.
5. Select "Demo Algorithms".
6. Select "Demo07_Subclassing.cs".

**Expected Result:** The Text box shows "AlgorithmLoader:success!".
**Actual Result:** The Text box shows "Line 1: CS0041 - Unexpected error writing debug information -- 'The version of Windows PDB writer is older than required: 'diasymreader.dll''".

**Findings:**
After comparing with Self-Contained app, there is a DLL named “Microsoft.DiaSymReader.Native.amd64.dll”, copy that file to Single file publish folder, launch the app again, it can load Algorithm successful.

Also, we find this assembly is displayed in TuringTrader\TuringTrader.Setup\PublishedFiles.wxs, so we think this assembly should be needed for the app.
But it is not included when we publish SingleFile. (The dll is included in SCD but not in SingleFile)

CC @dotnet-actwx-bot

**Comment by vitek-karas** [comment]
After a long trip into the inner workings of our shared library build system (across the SDK, Arcade, runtime repos) I finally figured this out.

Unlike in 3.1 where we simply packaged everything into the single-file, in 5.0 there's a list of native libraries from the framework which are supposed to be included in the single-file apps - everything else is removed. This list is defined here: https://github.com/dotnet/runtime/blob/ef6c035c399ba856566a9e3f4531084c8e87496a/src/installer/pkg/projects/netcoreapp/pkg/Directory.Build.props#L15-L32

As you can see the `Microsoft.DiaSymReader` library is not in the list - so it gets dropped from the single-file. 

In NETCoreApp the diasym reader is only used in debugger related functionality (as far as I can tell), so that's probably why it was removed from single-file (to reduce size). The problem is that Roslyn seems to rely on it being there. You can see that on .NET Framework Roslyn explicitly includes it in its packages
https://github.com/dotnet/roslyn/blob/e704ca635bd6de70a0250e34c4567c7a28fa9f6d/src/NuGet/Microsoft.Net.Compilers.Toolset/DesktopCompilerArtifacts.targets#L32-L33

but for .NET Core it doesn't.

The app in this issue uses `Microsoft.CodeAnalysis.dll` (from Roslyn) which uses the diasym reader - and I guess it assumes it's present in the framework.

**Comment by vitek-karas** [comment]
In short - Roslyn (as an example) depends on runtime shipping a certain native binary. If we think that runtime shipping DiaSym reader is a public feature (as in runtime expects others to take dependency on it), then there should be a way to declare this dependency. Unfortunately NuGet can't really do that today (nor the app if it was a direct dependency from the app).

Note: If we say this is not a public feature and we would expect Roslyn to ship DiaSym reader included, then there might be runtime issues if Roslyn loads its version and runtime tries to load its version - native library in-proc-SxS is relatively tricky and PInvokes/runtime currently doesn't support it really.

One solution would be to introduce a simple way to declare such dependency - nothing fancy, just MSBuild property - something like `<IncludeDiaSymReader>true</IncludeDiaSymReader>`. Roslyn could define this in its `.props` file. SDK would either drop or include the necessary files from single-files (and potentially also from self-contained non-single-file apps).

Eventually Roslyn could even implement this as a [feature switch](https://github.com/dotnet/designs/blob/main/accepted/2020/feature-switch.md) and provide some workaround if the library is not present.

**Comment by tommcdon** [comment]
FYI @noahfalk @mikem8361

**Comment by Junjun-zhao** [comment]
@agocke @LakshanF @vitek-karas  Verified with dotnet-sdk-6.0.100-preview.5.21266.3. still the same issue. Is there any update for this?

**Comment by agocke** [comment]
@Junjun-zhao No, I think we're willing to say this scenario is unsupported for now. This library is only used for writing Native PDBs, which aren't supported on any platform aside from Windows. I don't think we should be shipping this inside the shared framework at all.

The right fix for Roslyn would be to include this library in the `runtimes` directory for Windows.

**Comment by agocke** [comment]
Re-opening until I can find the right replacement bug to open for DiaSymReader native

**Comment by Junjun-zhao** [comment]
@agocke  Thanks for your update!

**Comment by VSadov** [comment]
@agocke - I assume this either need to be moved out of 6.0 or closed, if there is a right replacement bug.
